worlds/hk/__init__.py

```diff
@@ -165,75 +167,161 @@ def white_palace_exclusions(self):
             exclusions.update(path_of_pain_locations)
         if wp <= WhitePalace.option_kingfragment:
             exclusions.update(white_palace_checks)
-        if wp == WhitePalace.option_exclude and self.world.RandomizeCharms[self.player]:
-            # Ensure KF location is still reachable if charms are non-randomized
-            exclusions.update(white_palace_transitions)
-            exclusions.update(white_palace_events)
+        if wp == WhitePalace.option_exclude:
             exclusions.add("King_Fragment")
+            if self.world.RandomizeCharms[self.player]:
+                # If charms are randomized, this will be junk-filled -- so transitions and events are not progression
+                exclusions.update(white_palace_transitions)
+                exclusions.update(white_palace_events)
         return exclusions
 
     def create_regions(self):
         menu_region: Region = create_region(self.world, self.player, 'Menu')
         self.world.regions.append(menu_region)
-        wp_exclusions = self.white_palace_exclusions()
+        # wp_exclusions = self.white_palace_exclusions()
 
         # Link regions
         for event_name in event_names:
-            if event_name in wp_exclusions:
-                continue
+            #if event_name in wp_exclusions:
+            #    continue
             loc = HKLocation(self.player, event_name, None, menu_region)
             loc.place_locked_item(HKItem(event_name,
-                                         event_name not in wp_exclusions,
+                                         True, #event_name not in wp_exclusions,
                                          None, "Event", self.player))
             menu_region.locations.append(loc)
         for entry_transition, exit_transition in connectors.items():
-            if entry_transition in wp_exclusions:
-                continue
+            #if entry_transition in wp_exclusions:
+            #    continue
             if exit_transition:
                 # if door logic fulfilled -> award vanilla target as event
                 loc = HKLocation(self.player, entry_transition, None, menu_region)
                 loc.place_locked_item(HKItem(exit_transition,
-                                             exit_transition not in wp_exclusions,
+                                             True, #exit_transition not in wp_exclusions,
                                              None, "Event", self.player))
                 menu_region.locations.append(loc)
 
+    def create_shop_item(self, needed_extra_shop_slots):
```

Comment:
I'd ditch this entirely.  The items placed here aren't logically associated with the shops so this seems like the wrong place to place them (move them to `create_items()` instead

To balance item vs location count, we can do the following at the very end of create_items().  (This isn't tested, btw)

```python
        item_count = len(self.world.itempool)
        location_count = len(self.world.get_unfilled_locations())
        if location_count < item_count:  # More items than locations.
            # We only add at maximum 5 extra items.  Even if they all get assigned to the same shop, it is not possible
            # to exceed the 16 slots per shop limit here.  This code will need to revised if significantly more extra
            # items get added in the future.
            shops = list(multi_locations.keys)
            if not self.world.EggShopSlots[self.player].value:  # No eggshop, so don't place items there
                shops.remove('Egg_Shop')
            while location_count < item_count:
                # Add a location to a random shop.
                self.create_location(self.world.random.choice(shops))
                location_count += 1
        elif item_count < location_count:  # More locations than items
            for _ in range(location_count - item_count):
                # Create enough filler items to fill all locations.
                self.create_item(self.get_filler_item_name())
```